chain/client/src/sync/state.rs

```diff
@@ -800,6 +727,245 @@ impl StateSync {
             _ => {}
         }
     }
+
+    /// Checks if the header is downloaded.
+    /// If the download is complete, then moves forward to StateDownloadParts, otherwise retries the header request.
+    /// Returns (download_timeout, need_shard).
+    fn sync_shards_download_header_status(
+        &mut self,
+        shard_id: ShardId,
+        shard_sync_download: &mut ShardSyncDownload,
+        sync_hash: CryptoHash,
+        chain: &mut Chain,
+        now: DateTime<Utc>,
+    ) -> Result<(bool, bool), near_chain::Error> {
+        let mut download_timeout = false;
+        let mut need_shard = false;
+        // StateDownloadHeader is the first step. We want to fetch the basic information about the state (its size, hash etc).
+        if shard_sync_download.downloads[0].done {
+            let shard_state_header = chain.get_state_header(shard_id.clone(), sync_hash)?;
+            let state_num_parts =
+                get_num_state_parts(shard_state_header.state_root_node().memory_usage);
+            // If the header was downloaded successfully - move to phase 2 (downloading parts).
+            // Create the vector with entry for each part.
+            *shard_sync_download = ShardSyncDownload {
+                downloads: vec![
+                    DownloadStatus {
+                        start_time: now,
+                        prev_update_time: now,
+                        run_me: Arc::new(AtomicBool::new(true)),
+                        error: false,
+                        done: false,
+                        state_requests_count: 0,
+                        last_target: None,
+                    };
+                    state_num_parts as usize
+                ],
+                status: ShardSyncStatus::StateDownloadParts,
+            };
+            need_shard = true;
+        } else {
+            let prev = shard_sync_download.downloads[0].prev_update_time;
+            let error = shard_sync_download.downloads[0].error;
+            download_timeout = now - prev > self.timeout;
+            // Retry in case of timeout or failure.
+            if download_timeout || error {
+                shard_sync_download.downloads[0].run_me.store(true, Ordering::SeqCst);
+                shard_sync_download.downloads[0].error = false;
+                shard_sync_download.downloads[0].prev_update_time = now;
+            }
+            if shard_sync_download.downloads[0].run_me.load(Ordering::SeqCst) {
+                need_shard = true;
+            }
+        }
+        Ok((download_timeout, need_shard))
+    }
+
+    /// Checks if the parts are downloaded.
+    /// If download of all parts is complete, then moves forward to StateDownloadScheduling.
+    /// Otherwise, retries the failed part downloads.
+    /// Returns (download_timeout, need_shard).
```

Comment:
same nit: instead of giving only the variable names, it could be extra nice to describe their semantics

Reply:
Done